### DISCUSSION TITLE
LoggerAwareInterface::setLogger() vs injecting LoggerInterface into the constructor

### DISCUSSION THREAD

**Discussion by webignition** [symfony/symfony#53921] [Q&A]
In preparing to add some logging to a service, I decided to read the [latest logging documentation](https://symfony.com/doc/current/logging.html#using-a-logger-inside-a-service) to check that I wasn't heading in the wrong direction.

The documentation suggests implementing `LoggerAwareInterface::setLogger()` as a means of using a logger inside a service.

I was expecting, as with other dependencies, to inject a `LoggerInterface` instance into the service constructor. I wasn't aware that implementing `LoggerAwareInterface::setLogger()` is also supported.

Implementing `LoggerAwareInterface::setLogger()` seems a little more verbose and allows the logger to be changed after the service has been instantiated which seems like a bit of unnecessary extra functionality. Injecting a `LoggerInterface` instance seems more in line with how service dependencies are otherwise handled and therefore a little more consistent.

Is there any advantage of implementing `LoggerAwareInterface::setLogger()`  over injecting a `LoggerInterface` instance into the constructor? Since the former is mentioned in the logging documentation, I can't help but wonder if there is a benefit to this approach that I'm failing to see.

**Comment by MatTheCat** [comment]
`LoggerAwareInterface` is part of [PSR-3](https://www.php-fig.org/psr/psr-3/#14-helper-classes-and-interfaces) so I guess there is a gain in term of interoperability if you’re developing a library :thinking: 

Note that the documentation also mention classic injection in [its first section](https://symfony.com/doc/current/logging.html#logging-a-message):

> To log a message, inject the default logger in your controller or service
